tests/federation_room_join_partial_state_test.go

```diff
@@ -188,6 +188,249 @@ func TestPartialStateJoin(t *testing.T) {
 		t.Logf("Alice sent event event ID %s", eventID)
 	})
 
+	// we should be able to receive typing EDU over federation during the resync
+	t.Run("CanReceiveTypingDuringPartialStateJoin", func(t *testing.T) {
+		deployment := Deploy(t, b.BlueprintAlice)
+		defer deployment.Destroy(t)
+		alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+		server := createTestServer(t, deployment)
+		cancel := server.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, server, alice.GetDefaultRoomVersion(t))
+		psjResult := beginPartialStateJoin(t, server, serverRoom, alice)
+		defer psjResult.Destroy()
+
+		derekUserId := psjResult.Server.UserID("derek")
+
+		content, _ := json.Marshal(map[string]interface{}{
+			"room_id": serverRoom.RoomID,
+			"user_id": derekUserId,
+			"typing":  true,
+		})
+		edu := gomatrixserverlib.EDU{
+			Type:    "m.typing",
+			Content: content,
+		}
+		psjResult.Server.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{}, []gomatrixserverlib.EDU{edu})
+
+		psjResult.FinishStateRequest()
+		alice.MustSyncUntil(t,
+			client.SyncReq{},
+			client.SyncEphemeralHas(serverRoom.RoomID, func(result gjson.Result) bool {
+				if result.Get("type").Str != "m.typing" {
+					return false
+				}
+				for _, item := range result.Get("content").Get("user_ids").Array() {
+					if item.Str == derekUserId {
+						return true
+					}
+				}
+				return false
```

Comment:
This checks that at least one user in the list of mxids is Derek. It'd be nice to also check that no other user is typing, i.e. that the list of typing users is exactly `[derekUserId]`.

Reply:
There may be a way to express this more neatly using complement's `must` library or the helpers in `internal/match/json.go`. Being blunt, I don't really understand the distinction between the two and neither seems intended to be used directly with a `gjson.Result`. Asking in the complement room or picking Kegan's brains might be worth a shot.

Reply:
I haven't looked yet at your pointers but I did learn since writing that code that I can use a json path in `Get`. Which is both nice and annoying since we are using `.` as a separator in types.
That plus your comment we should end up with a pretty simple equal expression here.

Reply:
I tweaked this a little bit in 39c553d.

I did look at `must`, but they mostly seemed to operate on byte arrays rather than `gjson.Result`s.